### PR TITLE
`history`: switch default `LocationState` generic from `any` to `unknown`

### PR DESCRIPTION
**Pull request by OliverJAsh**
`any` disables type checking, so it should be avoided:

```ts
declare const any: any;
any.foo.bar; // no compile error, but runtime exception!
```

Related:

- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41568
- https://github.com/microsoft/TypeScript/issues/36178
- https://github.com/DefinitelyTyped/DefinitelyTyped/issues/27012

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/history/index.d.ts [hunk 1]
```diff
@@ -49,7 +49,10 @@ export namespace History {
       location: Location<S>,
       action: Action,
     ) => void;
-    export type LocationState = any;
+    // The value type here is a "poor man's `unknown`". When these types support TypeScript
+    // 3.0+, we can replace this with `unknown`.
+    type PoorMansUnknown = {} | null | undefined;
```
Comment (danielnixon) [inline types/history/index.d.ts]: https://github.com/gcanti/unknown-ts/blob/master/src/index.ts#L1 maybe?

**Review by danielnixon — COMMENTED** [review]
I suspect this change would break a lot of code. Worth reading:
* https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25677#issuecomment-388215959

**Comment by OliverJAsh** [comment]
> I suspect this change would break a lot of code. Worth reading:

Perhaps we could wait until the next major version of `history` comes out, although it would be difficult to co-ordinate the releases so they're both released at exactly the same time.

File: types/history/index.d.ts (hunk 1 above)
Comment (OliverJAsh) [reply types/history/index.d.ts]: This has worked well for me so far, I've also used it in the Node types.

**Comment by johnnyreilly** [comment]
Thanks!

**Pull request merged**